iceoryx_examples/waitset/README.md

```diff
@@ -1,6 +1,566 @@
+|wait for triggers with timeout |`auto triggerVector = myWaitSet.timedWait(1_s);`  |
+|check if trigger originated from some class|`trigger.doesOriginateFrom(ptrToSomeClass)`|
+|get id of trigger|`trigger.getTriggerId()`|
+|call triggerCallback|`trigger()`|
+|acquire _TriggerOrigin|`trigger.getOrigin<OriginType>();`|
+|check if 2 trigger are logical equal|`trigger.isLogicalEqualTo(anotherTrigger)`|
+
+## Use cases
+This example consists of 5 use cases.
+ 
+ 1. `ice_waitset_gateway.cpp`: We build a gateway together to forward data
+    to another network. A list of subscribers is handled in an uniform way 
+    by defining a callback and just call it for every subscriber who received data.
+
+ 2. `ice_waitset_grouping`: We would like to group multiple subscribers into 2 
+    groups and handle them according to their group membership.
+
+ 3. `ice_waitset_individual`: A list of subscribers where every subscriber is 
+    handled differently.
+
+ 4. `ice_waitset_sync`: We use the WaitSet to trigger a cyclic call which should 
+    execute an algorithm every 100ms.
+
+ 5. `ice_waitset_trigger`: We create our own class which can be attached to a
+    WaitSet and trigger events.
+
+## Examples
+
+All our examples require a running `iox-roudi` and a running `iox-ex-waitset-publisher`
+which is sending the data. The publisher does not contain any _WaitSet_ specific
+logic and is explained in detail in the [icedelivery example](../icedelivery/).
+
+### Gateway
+
+We have a list of subscribers which can be subscribed to any arbitrary service
+and everytime we received a sample we would like to send the bytestream to a socket,
+write it into a file or print it to the console. But whatever we choose to do
+we perform the same task for all the subscribers.
+
+Lets start by implementing our callback which prints the subscriber pointer, the
+payload size and the payload pointer to the console.
+```cpp
+void subscriberCallback(iox::popo::UntypedSubscriber* const subscriber)
+{
+    subscriber->take().and_then([&](iox::popo::Sample<const void>& sample) {
+        std::cout << "subscriber: " << std::hex << subscriber << " length: " << std::dec
+                  << sample.getHeader()->m_info.m_payloadSize << " ptr: " << std::hex << sample.getHeader()->payload()
+                  << std::endl;
+    });
+}
+```
+A _Trigger_ always requires a callback which has the following signature 
+`void (TriggerOrigin)`. In our example the _TriggerOrigin_ is a
+`iox::popo::UntypedSubscriber*` which we use to acquire the latest sample which the subscriber
+has received (`take()`). When `take()` was successful we print our message to
+the console inside of the `and_then` lambda.
+
+In our `main` function we create a _WaitSet_ after we registered at our central
+broker RouDi and attach our `shutdownGuard` to it to handle `CTRL+c` events. 
+```cpp
+iox::popo::WaitSet waitset;
+
+shutdownGuard.attachToWaitset(waitset);
```

Comment:
What about renaming to `attach(iox::popo::WaitSet waitset)`? The signature has all the information already, only `iox::popo::WaitSet` can be attached with this method.

Reply:
I like this idea! @budrus @elBoberido what do you think. Let's say you can later have a `userTrigger` where you can attach a `WaitSet` and maybe also a threadsafe waitset which we call `Reactor` for now.

Should a user write:
```cpp
subscriber.attach(someWaitset);
subscriber.attach(someReactor);
//...
subscriber.detach();
```

Or should the user write
```cpp
subscriber.attachToWaitset(someWaitset);
subscriber.attachToReactor(someReactor);
//...
subscriber.detachWaitset();
subscriber.detachReactor();
```

I would prefer the first one with `attach` and `detach` since I do not see a benefit of a longer method name. It does not make the code more understandable.

Reply:
If you only call it attach, I would assume that you attach the waitset to the subscriber. But in fact it is the other way round

Reply:
Why not implementing it the other way round then? I could live with `attachTo(iox::popo::WaitSet waitset)`

Reply:
Would be better than only attach

Reply:
renamed to `attachTo` and the subscriber has the method `detachOf(SubscriberEvent)` to specify the event which should be detached.